### DISCUSSION TITLE
Deep Links doesn't work from internal browser

### DISCUSSION THREAD

**Discussion by RIP21** [ionic-team/capacitor#5211] [Q&A] [answered]
We have a redirect that is getting opened in the internal browser and the should redirect back and should be intercepted by a redirect link. It works fine for iOS, but for Android, for some reason, it sometimes works and sometimes doesn't.

We have `assetlinks.json` in `.well-known` folder. And it happens for local (I guess it's unsigned) and signed and published internal testing APK.

And we have intent filters set in the `AndroidManifest.xml` surely.

What can be the issue here?

```
      <intent-filter>
        <action android:name="android.intent.action.VIEW"/>

        <category android:name="android.intent.category.DEFAULT"/>
        <category android:name="android.intent.category.BROWSABLE"/>

        <data
          android:host="secured.community"
          android:scheme="https"/>
      </intent-filter>
```

[Working link with assetlinks.json](https://secured.community/.well-known/assetlinks.json)

And redirect is `https://secured.community/<some-redirect-url-here>` 

These are the versions we're using
```
"@capacitor/android": "^3.2.5",
    "@capacitor/app": "^1.0.5",
    "@capacitor/browser": "^1.0.5",
    "@capacitor/cli": "^3.2.5",
    "@capacitor/core": "^3.2.5",
    "@capacitor/ios": "^3.2.5",
    "@capacitor/local-notifications": "^1.0.8",
    "@capacitor/push-notifications": "^1.0.6",
```

**Comment by JarrodMFlesch** [comment]
Can you share how you are getting this to work with iOS? 

We are doing similar things and having a difficult time. We are using the @capacitor/Browser plugin to open the url in the app, and then attempting to redirect back to the app using deep links.

1. We placed the apple-app-site-association file on one of our servers under `/.well-known/apple-app-site-association`
2. Add domain to Xcode under associated domains in the form of `applinks:some-domain.com`
2. We open the external site url with the in-app browser
3. The external url will eventually redirect back to our hosted domain i.e. `some-domain.com/internal-app-path` (where the file in step 1 lives)
4. Attempt to catch that redirect with:
```javascript 
App.addListener('appUrlOpen', function (event: URLOpenListenerEvent) {
...
})
```

Are we missing a step? Interested to hear what you have done to get it working at all!

Thanks 🙂

**Reply by RIP21** [reply]
We do exactly the same for IOS and, apparently, it works. Make sure you have associated domains and other capabilities enabled in the Apple Developer site for this app.

I'm out of keyboard so I can't show you exact code. But it's pretty similar.

With Android is quite random and doesn't work reliably. For some phones it works for some it doesn't.

What's annoying tho is that no one engaged with me here for such a long time :( Capacitor is an amazing concept, but with these issues I'm unsure if it's worth it or just React Native is a better option. Browser is the only thing that been a huge pain in the ass that is not solved yet.

**Reply by RIP21** [reply]
@JarrodMFlesch if still intersted here is our code:
```typescript
export const DeepLinkListener = () => {
  const history = useHistory()
  useEffect(() => {
    App.addListener('appUrlOpen', (data) => {
      const slug = data.url.split(process.env.REACT_APP_HOST ?? '')[1]
      if (slug) {
        history.push(slug)
        void Browser.close() // This is needed for iOS, for Android it will close by itself after auth redirect interception
      }
    })
  }, [history])

  return null
}
```

**Reply by cyrilfr** [reply]
Chrome WebView blocks a deeplink if it's not explicitly triggered by a click of the user. How did you manage to trigger the deeplink programmatically?

**Comment by RIP21** [comment, marked as answer]
Ok, so the issue was with the `assetlinks.json` file that had the wrong SHA 256 in it. So for Android 11, it was working with suggesting opening in the app, but for Android 12 if you have the incorrect `assetlinks.json` your app will never intercept links unless they're trusted with `assetlinks.json` lookup. It will also never suggest opening a link with your app and Chrome let say, but would rather just open Chrome immediately.
In order to locally test this behavior, you need to go into App info and click into the `Open by default` section, and click "Add Link +" which will suggest links that are described in your `AndroidManifest` but are not verified. You need to go and explicitly enable them for testing one by one. But in prod if your SHA256 is correct they'll become verified once the app is updated/installed as lookup will be successful, so any URLs that are verified will be opened with your app and not with Chrome/Other Browser. 

And, yeah, the correct SHA 256 for a deployed app we found in the developer console of play market in the section with upload key or so.

Also don't forget to add `android:autoVerify="true` like `<intent-filter android:autoVerify="true">` to all URL/Deep Links/App Links entries.

So, long story short, if something is broken in that regard test `.well-known` files as mainly they're the issue :) Both for iOS and Android.
Then internal and external browsers will behave the same so you'll be able to use `Browser.open` for auth callback.

**Reply by shaffy9633** [reply]
Just to be clearer on 'In order to locally test this behavior, you need to go into App info and click into the Open by default section, and click "Add Link +" you need to go to ur emulator's device settings > Apps> Open By Default

**Reply by Huzefa3701** [reply]
I am facing the same issue, but in my case I am using web links (Deep Links) for deep linking in my Android app. Since I am using web links, I do not need to configure an `assetlinks.json` file.

The behavior is inconsistent — sometimes the redirect URL successfully opens the app, and sometimes it remains in the browser and does not redirect back to the app.

In this scenario, what could be the possible reasons for such intermittent behavior?
Also how should I resolve it?
